File: 0_1.baekjoon/shortest_path/java/src/BOJ2224.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class BOJ2224 {

    private static final int M = 'z' - 'A' + 1;
    private static final int BASE = 'A';
    private static boolean[][] connected;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());

        connected = new boolean[M][M];

        for (int i = 0; i < N; i++) {
            String line = br.readLine();
            int u = line.charAt(0) - BASE;
            int v = line.charAt(5) - BASE;

            connected[u][v] = true;
        }

        System.out.println(solve());

        br.close();
    }

    private static String solve() {
        StringBuilder sb = new StringBuilder();

        connect();

        int count = 0;
        for (int p = 0; p < M; p++) {
            for (int q = 0; q < M; q++) {
                if (p != q && connected[p][q]) {
                    char u = (char) (BASE + p);
                    char v = (char) (BASE + q);
                    sb.append(u).append(" => ").append(v).append("\n");
                    count++;

                }
            }
        }

        sb.insert(0, count + "\n");
        return sb.toString();
    }

    private static void connect() {
        for (int q = 0; q < M; q++) {
            for (int p = 0; p < M; p++) {
                for (int r = 0; r < M; r++) {
                    if (p == r) {
                        continue;
                    }
                    if (connected[p][r]) {
                        continue;
                    }
                    connected[p][r] = connected[p][q] && connected[q][r];
                }
            }
        }
    }
}

File: 0_1.baekjoon/shortest_path/kotlin/src/BOJ2224.kt
private const val BASE: Int = 'A'.code
private const val SIZE: Int = 'z'.code - BASE + 1
private lateinit var connected: Array<BooleanArray>
fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()
    connected = Array(SIZE) { BooleanArray(SIZE) }

    repeat(n) {
        val line = readLine()
        val p = line[0].code - BASE
        val q = line[5].code - BASE

        connected[p][q] = true
    }

    println(solve())
}

private fun solve(): String {
    connect()

    val sb = StringBuilder()
    var count = 0
    for (i in 0 until SIZE) {
        for (j in 0 until SIZE) {
            if (i == j || !connected[i][j]) continue
            val p = (BASE + i).toChar()
            val q = (BASE + j).toChar()
            sb.append(p).append(" => ").append(q).append("\n")
            count++
        }
    }
    sb.insert(0, count.toString() + "\n")
    return sb.toString()
}

private fun connect() {
    for (r in 0 until SIZE) {
        for (p in 0 until SIZE) {
            for (q in 0 until SIZE) {
                if (p == q || connected[p][q]) continue
                connected[p][q] = connected[p][r] && connected[r][q]
            }
        }
    }
}
